### PR TITLE
feat: implement FACT

### PR DESCRIPTION
**Pull request by ZikkeyLS**

### CODE REVIEW

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp [hunk 1]
```diff
@@ -305,6 +305,117 @@ VarValue PapyrusActor::WornHasKeyword(VarValue self,
...
+    return VarValue(actor->IsInFaction(factionRec.rec->GetId()));
+  }
+  return VarValue(false);
+}
+
+VarValue PapyrusActor::GetFactions(VarValue self,
+                                   const std::vector<VarValue>& arguments)
+{
+  VarValue result = VarValue();
+  *result.pArray = std::vector<VarValue>();
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp:356]: attempt to dereference null pointer

result.pArray.reset instead

Comment (Pospelove) [reply skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp:356]: or parray = ...

File: unit/EspmTest.cpp [hunk 2]
```diff
@@ -137,6 +137,34 @@ TEST_CASE("Loads Conditions", "[espm]")
...
+  auto& br = l.GetBrowser();
+  espm::CompressedFieldsCache cache;
+
+  auto form = br.LookupById(0x00EB091);
+  REQUIRE(form.rec->GetType() == "FACT");
+  auto data = reinterpret_cast<const espm::FACT*>(form.rec)->GetData(cache);
+  REQUIRE(form.rec->GetEditorId(cache) ==
+          std::string("KhajiitCaravanFaction"));
+
+  REQUIRE((static_cast<uint32_t>(data.flags) &
```
Comment (Pospelove) [inline unit/EspmTest.cpp:151]: REQUIRE(static_cast<uint32_t>(data.flags) == ...)

check all flags plz

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusSkymp.cpp [hunk 3]
```diff
@@ -13,3 +13,32 @@ VarValue PapyrusSkymp::SetDefaultActor(VarValue self,
 
   return VarValue::None();
 }
+
+VarValue PapyrusSkymp::GetFaction(VarValue self,
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusSkymp.cpp:17]: I don't know what's the best way to do this. let's put out of this PR for now, to make PR pass as fast as possible

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusFaction.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+VarValue PapyrusFaction::GetReaction(VarValue self,
+                                     const std::vector<VarValue>& arguments)
+{
+  if (arguments.empty()) {
+    spdlog::error("Faction.GetReaction - at least one argument expected");
+    return VarValue(-1);
+  }
+
+  if (arguments[0].GetType() != VarValue::kType_Object) {
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusFaction.cpp:11]: check not needed, already covered by GetRecordPtr

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusFaction.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  WorldState* worldState = compatibilityPolicy->GetWorldState();
+
+  espm::FACT::Data otherFactionData =
+    espm::Convert<espm::FACT>(otherFaction.rec)
+      ->GetData(worldState->GetEspmCache());
+
+  int reaction = -1;
+  for (const auto& interfactionRelation :
+       otherFactionData.interfactionRelations) {
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusFaction.cpp:27]: this is not our reaction. this is _their_ reaction to us

I guess this is a bug

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp [hunk 6]
```diff
@@ -305,6 +305,117 @@ VarValue PapyrusActor::WornHasKeyword(VarValue self,
...
+
+    return VarValue(actor->IsInFaction(factionRec.rec->GetId()));
+  }
+  return VarValue(false);
+}
+
+VarValue PapyrusActor::GetFactions(VarValue self,
+                                   const std::vector<VarValue>& arguments)
+{
+  VarValue result = VarValue();
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp:355]: Type must be set

example:
https://github.com/skyrim-multiplayer/skymp/blob/4ae34fc81146d6211be550279a2fae4e409d603f/unit/VarValueTest.cpp#L258

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp [hunk 7]
```diff
@@ -305,6 +305,117 @@ VarValue PapyrusActor::WornHasKeyword(VarValue self,
...
+      throw std::runtime_error("Actor.GetFactions requires two arguments");
+    }
+
+    auto minFactionRank = static_cast<int>(arguments[0]);
+    auto maxFactionRank = static_cast<int>(arguments[1]);
+
+    if (minFactionRank < -128 || minFactionRank > 127 ||
+        maxFactionRank < -128 || maxFactionRank > 127 ||
+        minFactionRank > maxFactionRank) {
+      return result;
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp:374]: spdlog::warn in this case

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp [hunk 8]
```diff
@@ -305,6 +305,117 @@ VarValue PapyrusActor::WornHasKeyword(VarValue self,
...
+      throw std::runtime_error("Actor.IsInFaction requires one argument");
+    }
+
+    const auto& factionRec = GetRecordPtr(arguments[0]);
+    if (!factionRec.rec) {
+      spdlog::error("Actor.IsInFaction - invalid faction form");
+      return VarValue(false);
+    }
+
+    return VarValue(actor->IsInFaction(factionRec.rec->GetId()));
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp:347]: toGlobalId

File: skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp [hunk 9]
```diff
@@ -305,6 +305,117 @@ VarValue PapyrusActor::WornHasKeyword(VarValue self,
...
+    const auto& factionRec = GetRecordPtr(arguments[0]);
+    if (!factionRec.rec) {
+      spdlog::error("Actor.AddToFaction - invalid faction form");
+      return VarValue();
+    }
+
+    WorldState* worldState = compatibilityPolicy->GetWorldState();
+
+    Faction resultFaction = Faction();
+    resultFaction.formId = factionRec.rec->GetId();
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/script_classes/PapyrusActor.cpp:325]: toGlobalId

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 10]
```diff
@@ -581,15 +581,30 @@ bool WorldState::AttachEspmRecord(const espm::CombineBrowser& br,
...
+
+  MpChangeFormREFR* changeForm = nullptr;
   if (!isNpc) {
     form.reset(new MpObjectReference(formLocationalData,
                                      formCallbacksFactory(), baseId,
                                      typeStr.data(), primitiveBoundsDiv2));
   } else {
-    form.reset(
-      new MpActor(formLocationalData, formCallbacksFactory(), baseId));
+    MpActor* actor =
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:591]: Please move this logic to a better place. Consider lazy loading: create a separate method for that and use only on attempt to get/set factions.

Also, we need to respect NPC templates. see example in MpActor::GetRaceId

File: skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp [hunk 11]
```diff
@@ -908,6 +908,56 @@ void MpObjectReference::RegisterPrivateIndexedProperty(
   }
 }
 
+void MpObjectReference::AddToFaction(Faction faction)
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp:911]: In Skyrim ObjectReferences can't have faction ranks. Only Actors can. 

By the way AddToFaction is a good place to lazily read factions from NPC_ record

File: skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp [hunk 12]
```diff
@@ -908,6 +908,56 @@ void MpObjectReference::RegisterPrivateIndexedProperty(
...
+}
+
+void MpObjectReference::RemoveFromFaction(uint32_t factionFormID)
+{
+  EditChangeForm([&](MpChangeFormREFR& changeForm) {
+    if (!changeForm.factions.has_value()) {
+      return;
+    }
+
+    for (auto it = changeForm.factions.value().begin();
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/MpObjectReference.cpp:951]: Better remove all. There can be duplicates of some one does that in database by mistake.

File: skymp5-server/cpp/server_guest_lib/Faction.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+#include <simdjson.h>
+
+struct Faction
+{
+  uint32_t formId;
+  int8_t rank;
+
+  friend bool operator==(const Faction& r, const Faction& l)
+  {
+    return r.formId == l.formId;
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/Faction.h:12]: r.rank == l.rank

File: skymp5-server/cpp/server_guest_lib/Faction.h [hunk 14]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+#include <nlohmann/json.hpp>
+#include <simdjson.h>
+
+struct Faction
+{
+  uint32_t formId;
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/Faction.h:7]: = 0

File: libespm/include/libespm/FACT.h [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  {
+    uint32_t rankId = 0;
+    std::optional<uint32_t> maleTitle;
+    std::optional<uint32_t> femaleTitle;
+  };
+
+  struct Data
+  {
+    // fullName is lstring = uint - id in String Table (4 byte length)
+    std::optional<uint32_t> fullNameTableID;
```
Comment (Pospelove) [inline libespm/include/libespm/FACT.h:72]: we usually don't do optionals in espm as far as i know

**Review by Pospelove — CHANGES_REQUESTED** [review]
Good starting point, however some changes needed.

Also in the follow up PRs would be nice to bring factions to live on the client.

File: skymp5-server/cpp/server_guest_lib/Faction.h (hunk 13 above)
Comment (ZikkeyLS) [reply skymp5-server/cpp/server_guest_lib/Faction.h:12]: Why? I just want to check, whether the faction the same in its id.

Comment (Pospelove) [reply skymp5-server/cpp/server_guest_lib/Faction.h:12]: answered in dm

File: skymp5-server/cpp/server_guest_lib/MpActor.cpp [hunk 16]
```diff
@@ -220,6 +220,66 @@ void MpActor::SetEquipment(const std::string& jsonString)
...
+    LoadFactions();
+
+  EditChangeForm([&](MpChangeFormREFR& changeForm) {
+    if (!changeForm.factions.has_value()) {
+      return;
+    }
+
+    auto& factions = changeForm.factions.value();
+
+    for (int i = 0; i < factions.size(); i++) {
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/MpActor.cpp:274]: I think we should use standard algorithm here to make it obviously safe

```c++
factions.erase(
      std::remove_if(factions.begin(), factions.end(), 
                     [&](const Faction& faction) {
                       return faction.formId == factionFormID;
                     }),
      factions.end()
    );
    ```

File: skymp5-server/cpp/server_guest_lib/MpActor.cpp [hunk 17]
```diff
@@ -865,6 +925,26 @@ void MpActor::AddDeathItem()
...
 
+void MpActor::LoadFactions()
+{
+  std::vector<Faction> factions = EvaluateTemplate<espm::NPC_::UseFactions>(
+    GetParent(), GetBaseId(), GetTemplateChain(),
+    [](const auto& npcLookupResult, const auto& npcData) {
+      std::vector<Faction> factions = std::vector<Faction>();
+      for (auto npcFaction : npcData.factions) {
+        Faction faction = Faction();
+        faction.formId = npcFaction.formId;
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/MpActor.cpp:936]: npcLookupResult.toGlobalId

File: skymp5-server/cpp/server_guest_lib/MpActor.cpp [hunk 18]
```diff
@@ -865,6 +925,26 @@ void MpActor::AddDeathItem()
...
+        faction.rank = npcFaction.rank;
+        factions.push_back(faction);
+      }
+      return factions;
+    });
+  for (Faction faction : factions) {
+    AddToFaction(faction);
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/MpActor.cpp:943]: recursion

AddToFaction
LoadFaction
AddToFaction
LoadFaction
AddToFaction
LoadFaction
AddToFaction
LoadFaction
...

**Review by Pospelove — CHANGES_REQUESTED** [review]

**Review by Pospelove — APPROVED** [review]

**Pull request merged**